### ISSUE TITLE
Fingerprinting protections bypassable

### ISSUE THREAD

**Issue by snyderp** [brave/browser-laptop#11683]
## Test plan

----

### Description
Many of the current fingerprinting protections rely on removing references to methods from the global object.  However, you can get alternate references to the same methods by inserting a configured iframe element and grabbing the references off `iframe.contentWindow` or `iframe.contentDocument`.

### Steps to Reproduce
```
//blocked
console.log(HTMLCanvasElement.prototype.toDataURL);

let iframe = document.createElement("iframe");
iframe.src = "https://www.brave.com/";
document.body.appendChild(iframe);

// Not blocked
console.log(iframe.contentWindow.HTMLCanvasElement.prototype.toDataURL);

// Do the bad stuff
let canvasToDataUrl = iframe.contentWindow.HTMLCanvasElement.prototype.toDataURL;
canvasToDataUrl.apply(someParentFrameCanvasElement);
```

**Actual result:**
You can get references to the fingerprinting related methods.

**Expected result:**
You shouldn't be able to get to these methods.

**Reproduces how often:** 100%

### Brave Version
All versions

**Reproducible on current live release:**
Yes

**Comment by snyderp** [comment]
@diracdeltas Re our conversation on slack.

Happy to contribute a PR, but wanted to make sure my proposed solution was useful first.  How does overwriting the getter for the following methods to return the blocking proxy sound?

`HTMLIFrameElement.prototype.contentWindow`
`HTMLIFrameElement.prototype.contentDocument`
`HTMLFrameElement.prototype.contentWindow`
`HTMLFrameElement.prototype.contentDocument`

This will break some patterns of cross domain access, but (hopefully?) this is very rare, and most of the world has moved on to "Channel Messaging" for these use cases.

**Comment by diracdeltas** [comment]
apparently i tried to address this attack vector in https://github.com/brave/browser-laptop/commit/7e5541660f30d49378013db4b277e4b5756d02f5#diff-7440f7b95ccc371bdd73dce0a6631831R928

**Comment by diracdeltas** [comment]
nvm, the attack vector above is somewhat different since this issue is for using prototype methods in the child frames to do fingerprinting in the parent frame. @snyderp your approach sounds good to me.

**Comment by diracdeltas** [comment]
QA steps: go to https://jsfiddle.net/3dhmjqco/1/ with fingerprinting protection on, it should show 1 fingerprinting method blocked.

**Comment by kjozwiak** [comment]
@srirambv @LaurenWags this needs to be checked as well.. @diracdeltas has provided QA instructions in https://github.com/brave/browser-laptop/issues/11683#issuecomment-340532797.

**Comment by kjozwiak** [comment]
Before going through the verification, I reproduced the original issue using the STR from https://github.com/brave/browser-laptop/issues/11683#issuecomment-340532797 using `Brave: 0.19.70` rev: d4b94c6215bb130250fe5245787a5884dc9b51fe which is the current release build.

**_Verification:_**

* Ubuntu 17.04 x64 - PASSED (blocked the method when fingerprinting was enabled)

**Comment by diracdeltas** [comment]
@darkdh reports that on 0.19.75, gmail (mail.google.com) shows 1 canvas fingerprinting method blocked whereas before it would show 0. could it be that gmail was using this method for bypassing fp? i can't think of any legit reason to call `iframe.contentWindow.HTMLCanvasElement.prototype`

either that or it's a false positive

**Comment by srirambv** [comment]
@diracdeltas confirmed, with fingerprint protection enabled, gmail doesn't login gets stuck in loading progress bar. Have to disable to load inbox.

**Comment by kjozwiak** [comment]
Yup, seeing the same thing.... Checking the shield, the following method is being listed:
`Canvas: https://mail.google.com/mail/u/0/#inbox`

@diracdelta, do we know if this is expected or a false positive? We should find the answer to that question before shipping this with the hotfix.. This will break a lot of users experience with Gmail if they have FP enabled.

**Comment by snyderp** [comment]
Ah, there are two issues here.  One is that logging on really does break if you deny first person tracking on Google.  This is part of the "fix", i'm short of ideas on what to do about this at the moment, without either 1) loosing all finger printing protections, or 2) patching chromium to guarantee that content scripts are injected into frames before control is returned to the parent frame

The second half of this is a implementation bug in my patch (sorry!) that is fixed in 008a704a626b6c7bc541dc0514d765e93bf23a51

**Comment by snyderp** [comment]
Ok, the second issue (the false positive, not the breaking google login) is fixed with 05dcf1f

About breaking google the google login, im not sure how to square
  1) Don't break google
  2) Still block fingerprinting methods revealed on injected frames before `document_start` fires on in the child frame…

**Comment by diracdelta** [comment]
I began receiving messages from this thread. I think you have the wrong "diracdelta".

**Comment by diracdeltas** [comment]
@snyderp do you have any idea why Google is accessing canvas properties on an iframe in order to do the login? can you copy-paste the code where this happens?

**Comment by diracdeltas** [comment]
I think we should revert this change for now. Just realized it will probably break many sites because contentWindow is the recommended way to do postMessage on an iframe, ex: https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage

**Comment by snyderp** [comment]
@diracdeltas 

On more debugging of this specific case.  The issue here is unrelated to blocking the contentWindow.  On reading through google's packed source, the issue boils down to these lines of code

```
T9b.prototype.ra={
    for (var a=0;a<this.ha.length;a++) {
        var b = Fs(this.ha[a]).body;
        b&&ts(b)
    }
    this.ha.length=0;
    this.Gf()
}


b = function (a){for(var b;b=a.firstChild;)a.removeChild(b)}
```

in the above `b` is called with the proxy object (which was set on the contentWindow property on the frame), and the b function gets stuck in an infinite loop.

I think the best option here, and to prevent issues with infinite loops in general when stubbing out fingerprinting functionality, is to return `undefined` whenever the proxy has been accessed after some threshold (ex 1000 accesses).  This will cause an exception to be thrown whenever code gets into an infinite loop (like in this case).

I think this is the best of both worlds:
*  Prevents breaking pages that would bust if the finger printing methods were null'ed out (as before)
*  Prevent infinite loops that the proxy blocking technique could introduce in cases like this

e345e1a  implements this option

**Comment by luixxiul** [comment]
For Linux: https://github.com/brave/browser-laptop/pull/11784#discussion_r153139424

**Comment by LaurenWags** [comment]
@bsclifton @diracdeltas - Test Plan for this requires 'block all fingerprinting' which isn't available in 0.19.x but is in 0.20.x. should this issue be moved to 0.20.x?

cc @srirambv @kjozwiak

**Comment by bsclifton** [comment]
@LaurenWags good catch-  I don't know enough to answer that question

Deferring to @snyderp and @diracdelta

**Comment by LaurenWags** [comment]
thanks but most of the credit goes to @srirambv - will wait to see what @snyderp and @diracdeltas say regarding this issue.

**Comment by snyderp** [comment]
I would love to say "ship it" but I'm not really in a position to say :)  I think this is probably a 100% @diracdeltas question

**Comment by diracdeltas** [comment]
@LaurenWags 'block all fingerprinting' is the same as 'block fingerprinting' in 0.19.x

**Comment by LaurenWags** [comment]
so just to confirm @diracdeltas when test plan https://github.com/brave/browser-laptop/pull/11784#issuecomment-345803113 says FP count should be 1, but my Fingerprint count increases each time hangouts tries to open, this is expected?

![image](https://user-images.githubusercontent.com/28145373/33344390-de11dab2-d456-11e7-8df2-3616ab22e61b.png)

**Comment by LaurenWags** [comment]
Discussed https://github.com/brave/browser-laptop/issues/11683#issuecomment-347665009 with @diracdeltas and found that this is expected (but not ideal) behavior. Logged https://github.com/brave/browser-laptop/issues/12144 to address this.

**Comment by bsclifton** [comment]
PR has been reverted with https://github.com/brave/browser-laptop/commit/fe2fcedfdca94aeb29351fdd2388a3dca23fcedd

Re-opening so that we can look at a fix for 0.20.x

**Comment by bsclifton** [comment]
Moving to 0.21.x (since this would be fixed by https://github.com/brave/browser-laptop/issues/12045)
